guava/src/com/google/common/io/ReaderInputStream.java

```diff
@@ -185,6 +185,26 @@ public int read(byte[] b, int off, int len) throws IOException {
     }
   }
 
+  @Override
+  public int available() throws IOException {
+    if (byteBuffer.hasRemaining()) {
+      // if we have remaining chars in the buffer, return their count
+      return byteBuffer.remaining();
+    } else if (charBuffer.hasRemaining()) {
+      // if not, try to make quick progress by encoding available
+      // chars that have been read from the users reader
+      encoder.encode(charBuffer, byteBuffer, endOfInput);
```

Comment:
Would be good to check the result here, throwing an exception if the result indicates an error, or increasing the `charBuffer` or `byteBuffer` sizes as necessary:
- `charBuffer`: if `charBuffer` is full, but result is `UNDERFLOW` and `byteBuffer` is still empty
- `byteBuffer`: if result is `OVERFLOW` but `byteBuffer` is still empty; as far as I know there are some charset implementations which only but multiple bytes at bulk, and therefore would not put any bytes if the buffer has only size 1, which `ReaderInputStream` permits

`CharSource.asByteSource(...).openStream()` chooses buffer sizes where increasing the buffer should not be necessary, but the `ReaderInputStream` class is written in a style which would detect too small buffers and increases them as necessary, so it would be good to follow that.